modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/dml/DmlAstUtils.java

```diff
@@ -358,12 +362,72 @@ public static GridSqlSelect selectForUpdate(GridSqlUpdate update) {
 
         GridSqlElement where = update.where();
 
+        // On no MVCC mode we cannot use lazy mode when UPDATE query contains index with updated columns
+        // and that index may be chosen to scan by WHERE condition
+        // because in this case any rows update may be updated several times.
+        // e.g. in the cases below we cannot use lazy mode:
+        //
+        // 1. CREATE INDEX idx on test(val)
+        //    UPDATE test SET val = val + 1 WHERE val >= ?
+        //
+        // 2. CREATE INDEX idx on test(val0, val1)
+        //    UPDATE test SET val1 = val1 + 1 WHERE val0 >= ?
+        mapQry.canBeLazy(!isIndexWithUpdateColumnsMayBeUsed(
+            gridTbl,
+            update.cols().stream().map(sqlCol-> sqlCol.column()).collect(Collectors.toSet()),
+            extractColumns(gridTbl, where)));
+
         mapQry.where(where);
         mapQry.limit(update.limit());
 
         return mapQry;
     }
 
+    /**
+     * @return Set columns of the specified table that are used in expression.
+     */
+    private static Set<Column> extractColumns(GridH2Table tbl, GridSqlAst expr) {
+        if (expr == null)
+            return Collections.emptySet();
+
+        if (expr instanceof GridSqlColumn && ((GridSqlColumn)expr).column().getTable().equals(tbl))
+            return Collections.singleton(((GridSqlColumn)expr).column());
+
+        HashSet<Column> set = new HashSet<>();
+
+        for (int i = 0; i < expr.size(); ++i)
+            set.addAll(extractColumns(tbl, expr.child(i)));
+
+        return set;
+    }
+
+    /**
+     * @return {@code true} if the index contains update columns may be potentially used fot scan.
+     */
+    private static boolean isIndexWithUpdateColumnsMayBeUsed(
+        GridH2Table tbl,
+        Set<Column> updateCols,
+        Set<Column> whereCols) {
+        if (F.isEmpty(whereCols))
+            return false;
+
+        if (updateCols.size() == 1 && whereCols.size() ==1
+            && tbl.rowDescriptor().isValueColumn(F.first(updateCols).getColumnId())
+            && tbl.rowDescriptor().isValueColumn(F.first(whereCols).getColumnId()))
+            return true;
+
+        for (Index idx : tbl.getIndexes()) {
+            if (idx.equals(tbl.getPrimaryKey()) || whereCols.contains(idx.getColumns()[0])) {
```

Comment:
why you check only first index column? What about case `IDX1(col1, col2)` for `UPDATE tbl SET col2 = col2 + col1 WHERE col2 >=?`